lib/builddata.sh

```diff
@@ -1,29 +1,182 @@
+	build_data::set_raw "${key}" "${duration}"
+}
+
+# Sets a build data value as raw JSON data. The value parameter must be valid JSON value, that's also
+# a supported Honeycomb data type (string, integer, float, or boolean only; no arrays or objects).
+# For strings, use `build_data::set_string` instead since it will handle the escaping/quoting for you.
+# And for durations, use `build_data::set_duration`.
+#
+# Usage:
+# ```
+# build_data::set_raw "python_version_outdated" "true"
+# build_data::set_raw "foo_size_mb" "42.5"
+# ```
+function build_data::set_raw() {
+	local key="${1}"
+	local value="${2}"
+	build_data::_set "${key}" "${value}" "false"
+}
+
+# Internal helper to write a key/value pair to the build data store. The buildpack shouldn't call this directly.
+# Takes a key, value, and a boolean flag indicating whether the value needs to be quoted.
+#
+# Usage:
+# ```
+# build_data::_set "foo_string" "quote me" "true"
+# build_data::_set "bar_number" "99" "false"
+# ```
+function build_data::_set() {
+	local key="${1}"
+	# Truncate the value to an arbitrary 200 characters since it will sometimes contain user-provided
+	# inputs which may be unbounded in size. Ideally individual call sites will perform more aggressive
+	# truncation themselves based on the expected value size, however this is here as a fallback.
+	# (Honeycomb supports string fields up to 64KB in size, however, it's not worth filling up the
+	# build data store or bloating the payload passed back to Vacuole/submitted to Honeycomb given the
+	# extra content in those cases is not normally useful.)
+	local value="${2:0:200}"
+	local needs_quoting="${3}"
+
+	if [[ "${needs_quoting}" == "true" ]]; then
+		# Values passed using `--arg` are treated as strings, and so have double quotes added and any JSON
+		# special characters (such as newlines, carriage returns, double quotes, backslashes) are escaped.
+		local jq_args=(--arg value "${value}")
+	else
+		# Values passed using `--argjson` are treated as raw JSON values, and so aren't escaped or quoted.
+		local jq_args=(--argjson value "${value}")
+	fi
+
+	local new_data_file_contents
+	new_data_file_contents="$(jq --arg key "${key}" "${jq_args[@]}" '. + { ($key): ($value) }' "${BUILD_DATA_FILE}")"
+	echo "${new_data_file_contents}" >"${BUILD_DATA_FILE}"
+}
+
+# Retrieve the value of an entry in the build data store from the previous successful build.
+# Returns the empty string if the key wasn't found in the store.
+#
+# Usage:
+# ```
+# build_data::get_previous "python_version"
+# ```
+function build_data::get_previous() {
+	local key="${1}"
+
+	# Older versions of this buildpack used a `key=value` format file instead of JSON,
+	# so we need to support this file format/location too, so older caches can be read.
+	# We check for this file first, so that we correctly handle the case where an app
+	# downgraded and then re-upgraded buildpack version, so has both files in the cache.
+	if [[ -f "${LEGACY_BUILD_DATA_FILE}" ]]; then
```

Comment:
You'll need to delete the legacy data file at some point during the build, otherwise subsequent builds will continue to use this file rather than the newly written JSON file.

For Python I perform this deletion here (since I only use the previous build's values for cache invalidation, so once that handling has run, the legacy file is no longer needed):
https://github.com/heroku/heroku-buildpack-python/blob/30086f2e8da2bf50d1b669b2e9ca82d479bf0892/lib/cache.sh#L149-L152

And for Python this is covered by integration test that test upgrading from a previous buildpack version.

That said, I don't see `get_previous()` being called from anywhere at the moment? Is support for the legacy file even needed?

Reply:
Right now, no. But I was thinking of improving cache invalidation output at a later point, so I'd like to keep it. Do you think it's fine to leave it without the legacy file cleanup for now?